packages/core/src/browser/preferences/preference-proxy.ts

```diff
@@ -33,21 +33,103 @@ export interface PreferenceEventEmitter<T> {
     readonly ready: Promise<void>;
 }
 
+/**
+ * Generic interface to declare a typesafe get function based on the given
+ * configuration type.
+ *
+ * ### Example
+ *
+ * ```ts
+ * interface PreferenceConfiguration {
+ *  'myext.decorations.enabled': boolean,
+ *  'myext.debug': boolean,
+ *  'myext.editor.previewtype': string
+ * }
+ * const prefs : PreferenceRetrieval<PreferenceConfiguration> = createPreferenceProxy(...);
+ * const enabled : boolean = prefs.get('myext.decorations.enabled'); // valid
+ * const debug : string = prefs.get('myext.debug'); // invalid
+ * prefs.get('foobar'); // invalid
+ * ```
+ */
 export interface PreferenceRetrieval<T> {
     get<K extends keyof T>(preferenceName: K | {
         preferenceName: K,
         overrideIdentifier?: string
     }, defaultValue?: T[K], resourceUri?: string): T[K];
 }
 
+/**
+ * Typesafe schema-based preferences utility based on the {@link PreferenceService}.
+ *
+ * Can be used to set and get preferences as well as listen to preference changes.
+ *
+ * ### Example usage
+ *
+ * ``` ts
+ * const MyPreferencesSchema: PreferenceSchema = {
+ *   'type': 'object',
+ *   'properties': {
+ *       'myext.decorations.enabled': {
+ *           'type': 'boolean',
+ *           'description': 'Show file status',
+ *           'default': true
+ *       },
+ *       // [...]
+ *    }
+ * }
+ *
+ * interface MyPrefConfiguration {
+ *   'myext.decorations.enabled': boolean;
+ * }
+ *
+ * type MyPreferences = PreferenceProxy<MyPrefConfiguration>;
+ *
+ * const preferences : MyPreferences = createPreferenceProxy(preferenceService, MyPreferenceSchema);
+ *
+ * preferences.onPreferenceChanged(({ preferenceName, newValue }) => { console.log(preferenceName, newValue) });
+ * const enabled = preferences['myext.decorations.enabled'];
+ * ```
+ *
+ */
```

Comment:
Would it be better to link to an actual example in the codebase (ex: [core-preferences.ts](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/core-preferences.ts))?
My reasoning for this approach would be that:
- the actual example would be maintained and updated, while the example provided might get outdated.
- the actual example is syntax highlighted (helps the readability).
- the actual example describes how to bind it to the container which is not covered here.

I'm of course open to suggestions on whether we should keep it this way or not.

Reply:
I very much like the idea to link to an actual example. However I still think that this documentation here has some merit as it distills the usage down to what really matters in regards to the `PreferenceProxy`. I tried to not give a whole set of instructions of what the user can do but what they really need in practice, i.e. how to create and access the preferences. For me the binding was out of scope but I see how one can claim that it's missing.

Regarding the syntax highlighting: When hovering over the `PreferenceProxy` type the documentation shown in the popup is actually highlighted.

I'll reword it a bit and link to a real example.

Reply:
I removed the code example and replaced it with the general principle behind it while linking to the `CorePreferences` (also verified that the link works in the generated docs).